src/wallet/rpcwallet.cpp

```diff
@@ -198,20 +203,37 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
     //ret.pushKV("ischange", pwallet->IsChange(scriptPubKey));
 
     ScriptPubKeyMan* spk_man = pwallet->GetScriptPubKeyMan();
-    if (spk_man) {
-        CKeyID* keyID = boost::get<CKeyID>(&dest);
+    SaplingScriptPubKeyMan* sspk_man = pwalletMain->GetSaplingScriptPubKeyMan();
+    CKeyMetadata meta;
+
+    if (spk_man && pTransDest) {
+        // transparent destination
+        const CKeyID* keyID = boost::get<CKeyID>(pTransDest);
         if (keyID) {
-            std::map<CKeyID, CKeyMetadata>::iterator it = pwallet->mapKeyMetadata.find(*keyID);
+            auto it = pwallet->mapKeyMetadata.find(*keyID);
             if(it != pwallet->mapKeyMetadata.end()) {
-                const CKeyMetadata& meta = it->second;
-                ret.pushKV("timestamp", meta.nCreateTime);
-                if (meta.HasKeyOrigin()) {
-                    ret.pushKV("hdkeypath", meta.key_origin.pathToString());
-                    ret.pushKV("hdseedid", meta.hd_seed_id.GetHex());
-                    ret.pushKV("hdmasterfingerprint", HexStr(meta.key_origin.fingerprint, meta.key_origin.fingerprint + 4));
-                }
+                meta = it->second;
             }
         }
+    } else if (sspk_man && !pTransDest) {
+        // shield destination
+        const libzcash::SaplingPaymentAddress pa = *Standard::GetShieldedDestination(dest);
+        libzcash::SaplingExtendedSpendingKey extsk;
+        if (pwalletMain->GetSaplingExtendedSpendingKey(pa, extsk)) {
+            const auto& ivk = extsk.expsk.full_viewing_key().in_viewing_key();
+            auto it = sspk_man->mapSaplingZKeyMetadata.find(ivk);
+            if (it != sspk_man->mapSaplingZKeyMetadata.end()) {
+                meta = it->second;
+            }
+        }
+    }
+
+    // Add metadata
+    ret.pushKV("timestamp", meta.nCreateTime);
+    if (meta.HasKeyOrigin()) {
```

Comment:
The metadata could still be null here if there is no metadata for the address inside the wallet, and be loaded with the default constructor initialization (line 202). 
To keep the same flow as before, and not return the default values, should move it to a pointer and skip it if it's null.

Reply:
done